File: volume_1/chapter_4/web.py
"""
web.py - Clean web interface for LLM testers
Now supports optional memory endpoints and session-based tracking.
"""

from fastapi import Body, FastAPI, HTTPException
from fastapi.middleware.cors import CORSMiddleware
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
from typing import Optional, Union
import uvicorn
import io
from contextlib import redirect_stdout
import json
import os
import sys

REPO_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", ".."))
if REPO_ROOT not in sys.path:
    sys.path.append(REPO_ROOT)

from stream import iter_text_chunks, normalize_response_text
from utils import parse_structured_json_response, get_all_providers, get_default_model, get_display_name


class QueryRequest(BaseModel):
    topic: str
    provider: Optional[Union[str, int]] = None
    template: str = "{topic}"
    max_tokens: int = 1000
    temperature: float = 0.7
    session_id: Optional[str] = "default"
    sessionId: Optional[str] = None


class QueryAllRequest(BaseModel):
    topic: str
    template: str = "{topic}"
    max_tokens: int = 1000
    temperature: float = 0.7
    session_id: Optional[str] = "default"
    sessionId: Optional[str] = None


class ResetMemoryRequest(BaseModel):
    provider: Optional[Union[str, int]] = None
    session_id: Optional[str] = None
    sessionId: Optional[str] = None


def _supports_memory(manager) -> bool:
    """True when manager replays full chat history into prompts."""
    return bool(
        getattr(manager, "memory_enabled", False)
        and hasattr(manager, "get_history")
        and hasattr(manager, "reset_memory")
    )


def _supports_memory_retrieval(manager) -> bool:
    """True when manager supports retrieval-based memory context."""
    return bool(
        getattr(manager, "retrieval_memory_enabled", False)
        and hasattr(manager, "get_history")
        and hasattr(manager, "reset_memory")
    )


def _supports_session_memory(manager) -> bool:
    """True when any session-based memory mode (full replay or retrieval) is available."""
    return _supports_memory(manager) or _supports_memory_retrieval(manager)


def _supports_coagent(manager) -> bool:
    """True when manager exposes coagent features."""
    return bool(getattr(manager, "coagent", False))


def _parse_structured_raw_response(raw_response):
    if raw_response is None:
        return None
    try:
        parsed = parse_structured_json_response(raw_response)
        return parsed if isinstance(parsed, dict) else None
    except Exception:
        return None


def _extract_answer_text(response_obj, raw_response):
    if isinstance(response_obj, dict):
        answer = response_obj.get("answer") or response_obj.get("distilled") or response_obj.get("summary")
        if isinstance(answer, str) and answer.strip():
            return answer
    return normalize_response_text(raw_response)

def _recover_structured_parse_error(result: dict) -> dict:
    """Recover structured outputs without dropping fields like keywords."""
    if not isinstance(result, dict):
        return result

    raw_response = result.get("raw_response")
    parsed_response = _parse_structured_raw_response(raw_response)

    if result.get("success"):
        if parsed_response:
            response = result.get("response")
            metadata_notes = None
            if isinstance(response, dict):
                metadata_notes = (response.get("metadata") or {}).get("notes")
            if isinstance(metadata_notes, str) and metadata_notes.startswith("Failed to parse structured JSON response"):
                return {
                    **result,
                    "response": parsed_response,
                    "raw_answer": _extract_answer_text(parsed_response, raw_response),
                }
        return result

    error_message = str(result.get("error") or "")
    if "Failed to parse structured JSON response" not in error_message or not raw_response:
        return result

    recovered_response = parsed_response or {
        "answer": normalize_response_text(raw_response),
        "distilled": normalize_response_text(raw_response),
        "metadata": {
            "confidence": "low",
            "notes": error_message,
        },
    }

    return {
        **result,
        "success": True,
        "error": None,
        "response": recovered_response,
        "raw_answer": _extract_answer_text(recovered_response, raw_response),
    }


def _provider_selection_map(manager):
    """Build the same provider ordering used by the CLI selection prompt."""

    available = manager.get_available_providers()
    sorted_providers = sorted(available, key=lambda provider: (provider != "openai", get_display_name(provider)))
    return {str(index): provider for index, provider in enumerate(sorted_providers, start=1)}


def _normalize_provider_input(manager, provider: Optional[Union[str, int]]):
    """Accept provider numbers (CLI style) and provider keywords (web style)."""
    if provider is None:
        return None

    provider_map = _provider_selection_map(manager)
    available = {name.lower() for name in manager.get_available_providers()}
    configured = {name.lower() for name in get_all_providers()}

    if isinstance(provider, int):
        return provider_map.get(str(provider))

    candidate = str(provider).strip()
    if not candidate:
        return None

    if candidate in provider_map:
        return provider_map[candidate]

    lowered = candidate.lower()
    if lowered in available or lowered in configured:
        return lowered

    return candidate




def _result_value(result: dict, *keys, default=None):
    for key in keys:
        if isinstance(result, dict) and key in result and result.get(key) is not None:
            return result.get(key)
    return default




def _is_corrupt_history_error(exc: Exception) -> bool:
    message = str(exc)
    signals = (
        "string indices must be integers",
        "Expecting value",
        "JSON",
    )
    return any(signal in message for signal in signals)


def _ask_question_with_recovery(manager, args: dict, session_id: Optional[str]):
    """Run manager.ask_question with a one-time session-memory recovery retry."""
    try:
        return manager.ask_question(**args)
    except Exception as exc:
        if not _supports_session_memory(manager) or not _is_corrupt_history_error(exc):
            raise

        provider = args.get("provider")
        try:
            manager.reset_memory(provider, session_id)
        except Exception:
            raise exc

        return manager.ask_question(**args)


def create_web_api(manager_class):
    app = FastAPI(
        title="LLM Service API",
        version="1.0.0",
        description="Universal API for LLM framework testing"
    )

    app.add_middleware(
        CORSMiddleware,
        allow_origins=["*"],
        allow_credentials=True,
        allow_methods=["*"],
        allow_headers=["*"],
    )

    manager = manager_class()

    @app.get("/")
    async def get_status():
        available = manager.get_available_providers()
        return {
            "framework": manager.framework,
            "available_providers": available,
            "total_available": len(available),
            "initialization_status": manager.initialization_messages,
            "status": "healthy" if available else "no_providers"
        }

    @app.get("/providers")
    async def get_providers():
        providers = manager.get_available_providers()
        return {
            "framework": manager.framework,
            "providers": [
                {
                    "name": p,
                    "display_name": get_display_name(p),
                    "model": get_default_model(p),
                    "status": manager.initialization_messages.get(p, "Unknown")
                } for p in providers
            ],
            "count": len(providers)
        }

    @app.get("/capabilities")
    async def get_capabilities():
        return {
            "framework": manager.framework,
            "streaming": True,
            "memory": _supports_memory(manager),
            "memory_retrieval": _supports_memory_retrieval(manager),
            "coagent": _supports_coagent(manager),
        }

    @app.post("/query")
    async def query_single(request: QueryRequest):
        try:
            with redirect_stdout(io.StringIO()):
                args = {
                    "topic": request.topic,
                    "provider": _normalize_provider_input(manager, request.provider),
                    "template": request.template,
                    "max_tokens": request.max_tokens,
                    "temperature": request.temperature
                }
                effective_session_id = request.sessionId or request.session_id or "default"
                if _supports_session_memory(manager):
                    args["session_id"] = effective_session_id

                result = _recover_structured_parse_error(_ask_question_with_recovery(manager, args, effective_session_id))

            if not isinstance(result, dict):
                raise HTTPException(status_code=500, detail="Manager returned a non-dict response")

            if not result.get("success"):
                raise HTTPException(status_code=400, detail=result.get("error", "Query failed"))

            raw_response = result.get("response")
            content = raw_response if isinstance(raw_response, (dict, list)) else normalize_response_text(raw_response)

            return {
                "success": True,
                "framework": manager.framework,
                "provider": _result_value(result, "provider", default="unknown"),
                "model": _result_value(result, "model", default=""),
                "response": content,
                "parameters": {
                    "temperature": _result_value(result, "temperature"),
                    "max_tokens": _result_value(result, "max_tokens", "maxTokens"),
                    "template": request.template
                },
                "prompt": _result_value(result, "prompt", default=request.template.format(topic=request.topic)),
                "session_id": _result_value(result, "session_id", "sessionId", default=effective_session_id)
            }

        except HTTPException:
            raise
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @app.post("/query-stream")
    async def query_stream(request: QueryRequest):
        async def stream_events():
            try:
                with redirect_stdout(io.StringIO()):
                    args = {
                        "topic": request.topic,
                        "provider": _normalize_provider_input(manager, request.provider),
                        "template": request.template,
                        "max_tokens": request.max_tokens,
                        "temperature": request.temperature,
                    }
                    effective_session_id = request.sessionId or request.session_id or "default"
                    if _supports_session_memory(manager):
                        args["session_id"] = effective_session_id

                    result = _recover_structured_parse_error(_ask_question_with_recovery(manager, args, effective_session_id))

                if not isinstance(result, dict):
                    error_payload = {"type": "error", "error": "Manager returned a non-dict response"}
                    yield f"data: {json.dumps(error_payload)}\n\n"
                    return

                if not result.get("success"):
                    error_payload = {"type": "error", "error": result.get("error", "Query failed")}
                    yield f"data: {json.dumps(error_payload)}\n\n"
                    return

                raw_response = result.get("response")
                response_text = normalize_response_text(raw_response)
                async for chunk in iter_text_chunks(response_text, delay_seconds=0.03):
                    payload = {"type": "chunk", "content": chunk}
                    yield f"data: {json.dumps(payload)}\n\n"

                done_payload = {
                    "type": "done",
                    "provider": _result_value(result, "provider"),
                    "model": _result_value(result, "model"),
                    "response": raw_response if isinstance(raw_response, dict) else None,
                    "token_usage": _result_value(result, "token_usage", "tokenUsage"),
                    "session_id": _result_value(result, "session_id", "sessionId", default=effective_session_id),
                }
                yield f"data: {json.dumps(done_payload)}\n\n"
            except Exception as e:
                payload = {"type": "error", "error": str(e)}
                yield f"data: {json.dumps(payload)}\n\n"

        return StreamingResponse(stream_events(), media_type="text/event-stream")

    @app.post("/query-all")
    async def query_all(request: QueryAllRequest):
        try:
            with redirect_stdout(io.StringIO()):
                result = manager.query_all_providers(
                    topic=request.topic,
                    template=request.template,
                    max_tokens=request.max_tokens,
                    temperature=request.temperature
                )

            if not isinstance(result, dict):
                raise HTTPException(status_code=500, detail="Manager returned a non-dict response")

            if not result.get("success"):
                raise HTTPException(status_code=400, detail=result.get("error", "Query failed"))

            responses = result.get("responses")
            if not isinstance(responses, dict):
                raise HTTPException(status_code=500, detail="Manager returned invalid responses payload")

            clean_responses = {}
            for provider, res in responses.items():
                if not isinstance(res, dict):
                    clean_responses[provider] = {
                        "success": False,
                        "model": "",
                        "response": normalize_response_text(res),
                        "parameters": {
                            "temperature": None,
                            "max_tokens": None,
                        },
                    }
                    continue

                raw_content = _result_value(res, "response")
                content = raw_content if isinstance(raw_content, (dict, list)) else normalize_response_text(raw_content)
                clean_responses[provider] = {
                    "success": bool(_result_value(res, "success", default=False)),
                    "model": _result_value(res, "model", default=""),
                    "response": content,
                    "parameters": {
                        "temperature": _result_value(res, "temperature"),
                        "max_tokens": _result_value(res, "max_tokens", "maxTokens"),
                    }
                }

            return {
                "success": True,
                "framework": manager.framework,
                "prompt": _result_value(result, "prompt", default=request.template.format(topic=request.topic)),
                "responses": clean_responses
            }

        except HTTPException:
            raise
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @app.get("/history")
    async def get_history(provider: str = "openai", session_id: str = "default"):
        if not _supports_session_memory(manager):
            raise HTTPException(status_code=400, detail="Session memory not supported by this manager")
        return manager.get_history(provider, session_id)

    @app.post("/reset-memory")
    async def reset_memory(
        request: Optional[ResetMemoryRequest] = Body(None),
        provider: Optional[str] = None,
        session_id: Optional[str] = None,
    ):
        if not _supports_session_memory(manager):
            raise HTTPException(status_code=400, detail="Session memory not supported by this manager")
        body_provider = request.provider if request else None
        body_session_id = (request.sessionId or request.session_id) if request else None
        effective_provider = body_provider if body_provider is not None else provider
        effective_session_id = body_session_id if body_session_id is not None else session_id
        return manager.reset_memory(effective_provider, effective_session_id)

    return app


def run_web_server(manager_class, host: str = "0.0.0.0", port: int = 8000):
    app = create_web_api(manager_class)
    try:
        framework_name = manager_class().framework
    except Exception:
        framework_name = "Unknown"

    print(f"Starting web server for {framework_name}")
    print(f"Docs: http://{host}:{port}/docs")
    print(f"Health: http://{host}:{port}/")
    uvicorn.run(app, host=host, port=port)


def main():
    print("Universal LLM Web API")
    print("Run using `run_web_server(manager_class)`")


if __name__ == "__main__":
    main()

File: volume_1/chapter_4/web.js
/**
 * web.js - Clean web interface for JavaScript LLM testers.
 * Supports optional memory endpoints and session-based tracking.
 */

import { getAllProviders, getDefaultModelDetails, parseStructuredJsonResponse, sortProvidersByDisplayOrder } from './utils.js';
import { buildManager, captureConsoleOutputAsync, chunkText, createExpressApp, normalizeResponseText, resolveSessionId, resultIsSuccess, streamTextSse, supportsCoagent, supportsMemory, supportsMemoryRetrieval, supportsSessionMemory, toSseLine } from '../../shared/web.mjs';

function parseStructuredRawResponse(rawResponse) {
  if (typeof rawResponse === 'undefined' || rawResponse === null) return null;
  try {
    const parsed = parseStructuredJsonResponse(rawResponse);
    return parsed && typeof parsed === 'object' && !Array.isArray(parsed) ? parsed : null;
  } catch {
    return null;
  }
}

function extractAnswerText(responseObj, rawResponse) {
  if (responseObj && typeof responseObj === 'object') {
    const answer = responseObj.answer || responseObj.distilled || responseObj.summary;
    if (typeof answer === 'string' && answer.trim()) return answer;
  }
  return normalizeResponseText(rawResponse);
}

function recoverStructuredParseError(result) {
  if (!result) return result;
  const rawResponse = result.rawResponse;
  const parsedResponse = parseStructuredRawResponse(rawResponse);
  if (result.success) {
    if (parsedResponse && result.response && typeof result.response === 'object') {
      const metadataNotes = result.response?.metadata?.notes;
      if (typeof metadataNotes === 'string' && metadataNotes.startsWith('Failed to parse structured JSON response')) {
        return { ...result, response: parsedResponse, rawAnswer: extractAnswerText(parsedResponse, rawResponse) };
      }
    }
    return result;
  }
  const errorMessage = String(result.error || '');
  if (!errorMessage.includes('Failed to parse structured JSON response') || !rawResponse) return result;
  const normalized = normalizeResponseText(rawResponse);
  const recoveredResponse = parsedResponse || { answer: normalized, distilled: normalized, metadata: { confidence: 'low', notes: errorMessage } };
  return { ...result, success: true, error: null, response: recoveredResponse, rawAnswer: extractAnswerText(recoveredResponse, rawResponse) };
}

function providerSelectionMap(manager) {
  const sortedProviders = sortProvidersByDisplayOrder(manager.getAvailableProviders());
  return Object.fromEntries(sortedProviders.map((provider, index) => [String(index + 1), provider]));
}

function normalizeProviderInput(manager, provider) {
  if (provider === null || typeof provider === 'undefined') return null;
  const providerMap = providerSelectionMap(manager);
  const available = new Set(manager.getAvailableProviders().map((p) => String(p).toLowerCase()));
  const configured = new Set(getAllProviders().map((p) => String(p).toLowerCase()));
  if (typeof provider === 'number') return providerMap[String(provider)] ?? null;
  const candidate = String(provider).trim();
  if (!candidate) return null;
  if (candidate in providerMap) return providerMap[candidate];
  const lowered = candidate.toLowerCase();
  if (available.has(lowered) || configured.has(lowered)) return lowered;
  return candidate;
}

function buildProviderPayload(provider, manager) {
  const details = getDefaultModelDetails(provider);
  return {
    name: provider,
    display_name: details.displayName,
    provider: details.canonicalProvider,
    default_model: details.defaultModel,
    default_model_identifier: details.defaultModelIdentifier,
    default_model_tier: details.defaultModelTier,
    status: manager.initializationMessages[provider] || 'Unknown',
  };
}

export function createWebApi(managerClassOrFactory) {
  const app = createExpressApp();
  let manager;
  const initPromise = (async () => {
    manager = buildManager(managerClassOrFactory);
    await manager._checkProviders();
    return manager;
  })();
  app.use(async (_, __, next) => {
    if (!manager) await initPromise;
    next();
  });

  app.get('/', (_, res) => {
    const available = manager.getAvailableProviders();
    res.json({
      framework: manager.framework,
      available_providers: available,
      available_provider_details: available.map((provider) => buildProviderPayload(provider, manager)),
      total_available: available.length,
      initialization_status: manager.initializationMessages,
      status: available.length > 0 ? 'healthy' : 'no_providers',
    });
  });
  app.get('/providers', (_, res) => {
    const available = manager.getAvailableProviders();
    res.json({ framework: manager.framework, providers: available.map((provider) => buildProviderPayload(provider, manager)), count: available.length });
  });
  app.get('/capabilities', (_, res) => {
    res.json({ framework: manager.framework, streaming: true, memory: supportsMemory(manager), memory_retrieval: supportsMemoryRetrieval(manager), coagent: supportsCoagent(manager) });
  });

  app.post('/query', async (req, res) => {
    try {
      const { topic, provider = null, template = '{topic}', max_tokens = 1000, temperature = 0.7, session_id = 'default', sessionId = null } = req.body;
      if (!topic) return res.status(400).json({ error: 'Topic is required' });
      const effectiveSessionId = resolveSessionId(sessionId, session_id);
      const { result, logs } = await captureConsoleOutputAsync(async () => supportsSessionMemory(manager)
        ? recoverStructuredParseError(await manager.askQuestion(topic, normalizeProviderInput(manager, provider), template, max_tokens, temperature, effectiveSessionId))
        : recoverStructuredParseError(await manager.askQuestion(topic, normalizeProviderInput(manager, provider), template, max_tokens, temperature)));
      if (!resultIsSuccess(result)) return res.status(400).json({ error: result.error || 'Query failed', provider: result.provider, debug: logs || null });
      return res.json({ success: true, framework: manager.framework, provider: result.provider, model: result.model, response: (typeof result.response === 'object' && result.response !== null) ? result.response : normalizeResponseText(result.response), parameters: { temperature: result.temperature, max_tokens: result.maxTokens, template }, prompt: result.prompt, session_id: result.sessionId || effectiveSessionId, ...(logs ? { debug: logs } : {}) });
    } catch (error) {
      return res.status(500).json({ error: error.message, framework: manager.framework });
    }
  });

  app.post('/query-stream', async (req, res) => {
    try {
      const { topic, provider = null, template = '{topic}', max_tokens = 1000, temperature = 0.7, session_id = 'default', sessionId = null } = req.body;
      if (!topic) return res.status(400).json({ error: 'Topic is required' });
      const effectiveSessionId = resolveSessionId(sessionId, session_id);
      const { result } = await captureConsoleOutputAsync(async () => supportsSessionMemory(manager)
        ? recoverStructuredParseError(await manager.askQuestion(topic, normalizeProviderInput(manager, provider), template, max_tokens, temperature, effectiveSessionId))
        : recoverStructuredParseError(await manager.askQuestion(topic, normalizeProviderInput(manager, provider), template, max_tokens, temperature)));
      res.setHeader('Content-Type', 'text/event-stream');
      res.setHeader('Cache-Control', 'no-cache');
      res.setHeader('Connection', 'keep-alive');
      if (!resultIsSuccess(result)) {
        res.write(toSseLine({ type: 'error', error: result.error || 'Query failed' }));
        return res.end();
      }
      for await (const event of streamTextSse(normalizeResponseText(result.response), 28, 35)) {
        res.write(event);
      }
      res.write(toSseLine({ type: 'done', provider: result.provider, model: result.model, response: (typeof result.response === 'object' && result.response !== null) ? result.response : null, token_usage: result.tokenUsage ?? result.token_usage ?? null, session_id: result.sessionId ?? effectiveSessionId }));
      return res.end();
    } catch (error) {
      if (!res.headersSent) return res.status(500).json({ error: error.message, framework: manager.framework });
      res.write(toSseLine({ type: 'error', error: error.message }));
      return res.end();
    }
  });

  app.get('/history', async (req, res) => {
    if (!supportsSessionMemory(manager)) return res.status(400).json({ error: 'Session memory not supported by this manager' });
    const { provider = 'openai', session_id = 'default', sessionId = null } = req.query;
    return res.json(await Promise.resolve(manager.getHistory(normalizeProviderInput(manager, provider) || 'openai', sessionId ?? session_id ?? 'default')));
  });

  app.post('/reset-memory', async (req, res) => {
    if (!supportsSessionMemory(manager)) return res.status(400).json({ error: 'Session memory not supported by this manager' });
    const body = req.body || {};
    const provider = body.provider ?? req.query?.provider ?? null;
    const resetSessionId = body.sessionId ?? body.session_id ?? req.query?.sessionId ?? req.query?.session_id ?? null;
    return res.json(await Promise.resolve(manager.resetMemory(normalizeProviderInput(manager, provider), resetSessionId)));
  });

  app.get('/health', (_, res) => {
    const available = manager.getAvailableProviders();
    res.json({ status: available.length > 0 ? 'healthy' : 'unhealthy', framework: manager.framework, providers_available: available.length });
  });

  return app;
}

export async function runWebServer(managerClassOrFactory, host = '0.0.0.0', port = 8000) {
  const app = createWebApi(managerClassOrFactory);
  let frameworkName = 'Unknown';
  try { frameworkName = buildManager(managerClassOrFactory).framework; } catch {}
  app.listen(port, host, () => {
    console.log(`Starting web server for ${frameworkName} framework...`);
    console.log(`Health check: http://${host}:${port}/health`);
    console.log(`Status: http://${host}:${port}/`);
  });
}
